### PR TITLE
[2/2] Enable torch-xpu-ops on Native Windows

### PR DESCRIPTION
**Pull request by min-jean-cho**
As part of https://github.com/pytorch/pytorch/issues/126719, we are enabling torch-xpu-ops on Native Windows. 

The enabling is separated into two parts:

1. CMake changes to build on Windows https://github.com/intel/torch-xpu-ops/pull/287
2. Op changes to build on Windows https://github.com/intel/torch-xpu-ops/pull/288

This PR is the second part of the above. This PR:

- Fixes Windows or MSVC issues on some existing operator implementations. 

Currently, there's no CI/CD for Windows to gatekeep PRs that break on Windows. Hence this PR manually rebases to main, and fixes Window/MSVC issues from merged PRs. Once CI/CD is added for Windows, PRs will be gatekept to catch issues before merge.

### CODE REVIEW

File: src/aten/TensorFactories.cpp [hunk 1]
```diff
@@ -11,7 +11,7 @@
 #include <ATen/ops/empty_strided_native.h>
 #endif
 
-#include <aten/EmptyTensor.h>
+#include "EmptyTensor.h"
```
Comment (min-jean-cho) [inline src/aten/TensorFactories.cpp:14]: Issue due to case-insensitive on Windows vs. case-sensitive on Linux
See https://learn.microsoft.com/en-us/windows/wsl/case-sensitivity

aten/EmptyTensor.h: in torch-xpu-ops
ATen/EmptyTensor.h: in stock pytorch

- Case-sensitive, Linux: aten/EmptyTensor.h != ATen/EmptyTensor.h
- Case-insensitive, Windows: aten/EmptyTensor.h = ATen/EmptyTensor.h

Due to this, on Windows, aten/EmptyTensor.h is not being included properly, and errors missing XPU functions that are defined in aten/EmptyTensor.h.

**Comment by fengyuan14** [comment]
What's the plan of preci on Windows in torch-xpu-ops?

File: src/aten/TensorFactories.cpp (hunk 1 above)
Comment (fengyuan14) [reply src/aten/TensorFactories.cpp:15]: The file naming follows stock PyTorch. And implementation in the file might not be same. LGTM.

**Comment by min-jean-cho** [comment]
> What's the plan of preci on Windows in torch-xpu-ops?

I will update after sync offline with team.

**Comment by min-jean-cho** [comment]
Internal validation result:

- Version: https://github.com/intel/torch-xpu-ops/commit/c3c4d64e7c6b48b317e7cf1ec40b24f1a21bf55a
- Hardware: MTL
- OS: Windows 11
- UT scope: torch-xpu-ops/examples
- UT pass rate: 100%
examples_master

File: src/aten/sycl/Embedding.cpp [hunk 2]
```diff
@@ -47,8 +47,8 @@ Tensor embedding_dense_backward_kernel(
         AT_DISPATCH_INDEX_TYPES(
             indices.scalar_type(), "embedding_backward", [&] {
               // TODO: port pstl functions
-              auto sorted_begin = sorted_indices.data_ptr<index_t>();
-              auto orig_begin = orig_indices.data_ptr<index_t>();
+              index_t* sorted_begin = sorted_indices.data_ptr<index_t>();
+              index_t* orig_begin = orig_indices.data_ptr<index_t>();
```
Comment (min-jean-cho) [inline src/aten/sycl/Embedding.cpp:51]: Seem some MSVC peculiarity with type deduction.

File: src/aten/sycl/Indexing.cpp [hunk 3]
```diff
@@ -171,20 +171,24 @@ void index_select_kernel(
...
+          using IdxInfo = TensorInfo<index_t, int64_t>;
+
           // Improve efficiency of generated native instructions for contiguous.
           // See comm/TensorInfo.h
           if (dst.is_contiguous() && indices.is_contiguous())
             _index_select_kernel<
-                decltype(src_info),
-                decltype(dst_info),
-                decltype(index_info),
+                SrcInfo,
```
Comment (fengyuan14) [inline src/aten/sycl/Indexing.cpp:182]: Looks we need building test in Windows preci at least to gate compilation issues in latter development.

File: src/aten/sycl/SortingRadixSort.h [hunk 4]
```diff
@@ -2,6 +2,10 @@
 
 #include <aten/sycl/SortingCommon.h>
 
+#ifdef _WIN32
+#include <winsock.h>
```
Comment (fengyuan14) [inline src/aten/sycl/SortingRadixSort.h:6]: Which library call requires the header?

File: src/aten/sycl/Indexing.cpp (hunk 3 above)
Comment (min-jean-cho) [reply src/aten/sycl/Indexing.cpp:182]: Yes, definitely. cc. @ashokei

**Comment by min-jean-cho** [comment]
> What's the plan of preci on Windows in torch-xpu-ops?

> I will update after sync offline with team.

cc. @ashokei

File: src/aten/sycl/SortingRadixSort.h (hunk 4 above)
Comment (min-jean-cho) [reply src/aten/sycl/SortingRadixSort.h]: It was for `u_char` type in https://github.com/intel/torch-xpu-ops/blob/8cc6d5102878cefd3c71245fc7509a4548d07da2/src/aten/sycl/SortingRadixSort.h#L320

But I suppose `unsigned char` standard c++ type is better to use than `u_char`.

**Comment by min-jean-cho** [comment]
Closing due to too many merge conflicts after file structure modification. Moved to https://github.com/intel/torch-xpu-ops/pull/458.

**Pull request closed without merging**